File: WPF/Controls/NumericUpDown.vb
'Imports System.Collections.Specialized

Imports Ks.Common.MVVM

Namespace Common.Controls

    Public Class NumericUpDown
        Inherits Control

        Shared Sub New()
            DefaultStyleKeyProperty.OverrideMetadata(GetType(NumericUpDown), New FrameworkPropertyMetadata(GetType(NumericUpDown)))
        End Sub

#Region "Increment Command"
        Private _IncrementCommand As DelegateCommand = New DelegateCommand(AddressOf Me.OnIncrementCommand)

        Private Sub OnIncrementCommand()
            Me.Value += Me.Step
        End Sub

        Public ReadOnly Property IncrementCommand As DelegateCommand
            Get
                Return Me._IncrementCommand
            End Get
        End Property
#End Region

#Region "Decrement Command"
        Private _DecrementCommand As DelegateCommand = New DelegateCommand(AddressOf Me.OnDecrementCommand)

        Private Sub OnDecrementCommand()
            Me.Value -= Me.Step
        End Sub

        Public ReadOnly Property DecrementCommand As DelegateCommand
            Get
                Return Me._DecrementCommand
            End Get
        End Property
#End Region

#Region "Step Property"
        Public Shared ReadOnly StepProperty As DependencyProperty = DependencyProperty.Register("Step", GetType(Double), GetType(NumericUpDown), New PropertyMetadata(1.0, AddressOf Step_Changed, AddressOf Step_Coerce))

        Private Shared Function Step_Coerce(ByVal D As DependencyObject, ByVal BaseValue As Object) As Object
            Dim Self = DirectCast(D, NumericUpDown)

            Dim Value = DirectCast(BaseValue, Double)

            Return BaseValue
        End Function

        Private Shared Sub Step_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
            Dim Self = DirectCast(D, NumericUpDown)

            Dim OldValue = DirectCast(E.OldValue, Double)
            Dim NewValue = DirectCast(E.NewValue, Double)
        End Sub

        Public Property [Step] As Double
            Get
                Return DirectCast(Me.GetValue(StepProperty), Double)
            End Get
            Set(ByVal value As Double)
                Me.SetValue(StepProperty, value)
            End Set
        End Property
#End Region

#Region "Value Property"
        Public Shared ReadOnly ValueProperty As DependencyProperty = DependencyProperty.Register("Value", GetType(Double), GetType(NumericUpDown), New PropertyMetadata(0.0, AddressOf Value_Changed, AddressOf Value_Coerce))

        Private Shared Function Value_Coerce(ByVal D As DependencyObject, ByVal BaseValue As Object) As Object
            Dim Self = DirectCast(D, NumericUpDown)

            Dim Value = DirectCast(BaseValue, Double)

            Return BaseValue
        End Function

        Private Shared Sub Value_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
            Dim Self = DirectCast(D, NumericUpDown)

            Dim OldValue = DirectCast(E.OldValue, Double)
            Dim NewValue = DirectCast(E.NewValue, Double)
        End Sub

        Public Property Value As Double
            Get
                Return DirectCast(Me.GetValue(ValueProperty), Double)
            End Get
            Set(ByVal value As Double)
                Me.SetValue(ValueProperty, value)
            End Set
        End Property
#End Region

        Private IsDirty As Boolean

    End Class

End Namespace

File: WPF/Controls/NumericUpDown.cs
// Imports System.Collections.Specialized

using System.Windows;
using System.Windows.Controls;

using Ks.Common.MVVM;

namespace Ks.Common.Controls
{
    public class NumericUpDown : Control
    {
        static NumericUpDown()
        {
            DefaultStyleKeyProperty.OverrideMetadata(typeof(NumericUpDown), new FrameworkPropertyMetadata(typeof(NumericUpDown)));
        }

        public NumericUpDown()
        {
            this.IncrementCommand = new DelegateCommand(this.OnIncrementCommand);
            this.DecrementCommand = new DelegateCommand(this.OnDecrementCommand);
        }

        private void OnIncrementCommand()
        {
            this.Value += this.Step;
        }

        public DelegateCommand IncrementCommand { get; }

        private void OnDecrementCommand()
        {
            this.Value -= this.Step;
        }

        public DelegateCommand DecrementCommand { get; }

        public static readonly DependencyProperty StepProperty = DependencyProperty.Register("Step", typeof(double), typeof(NumericUpDown), new PropertyMetadata(1.0, Step_Changed, Step_Coerce));

        private static object Step_Coerce(DependencyObject D, object BaseValue)
        {
            var Self = (NumericUpDown) D;

            var Value = (double) BaseValue;

            return BaseValue;
        }

        private static void Step_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
        {
            var Self = (NumericUpDown) D;

            var OldValue = (double) E.OldValue;
            var NewValue = (double) E.NewValue;
        }

        public double Step
        {
            get => (double) this.GetValue(StepProperty);
            set => this.SetValue(StepProperty, value);
        }

        public static readonly DependencyProperty ValueProperty = DependencyProperty.Register("Value", typeof(double), typeof(NumericUpDown), new PropertyMetadata(0.0, Value_Changed, Value_Coerce));

        private static object Value_Coerce(DependencyObject D, object BaseValue)
        {
            var Self = (NumericUpDown) D;

            var Value = (double) BaseValue;

            return BaseValue;
        }

        private static void Value_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
        {
            var Self = (NumericUpDown) D;

            var OldValue = (double) E.OldValue;
            var NewValue = (double) E.NewValue;
        }

        public double Value
        {
            get => (double) this.GetValue(ValueProperty);
            set => this.SetValue(ValueProperty, value);
        }

#pragma warning disable CS0169 // The field is never used
        private bool IsDirty; // ToDo
#pragma warning restore CS0169 // The field is never used
    }
}
